### PR TITLE
BAEL-3298 Spring batch retry job

### PR DESCRIPTION
**Pull request by shubhi22**

### CODE REVIEW

File: spring-batch/src/test/java/org/baeldung/batch/SpringBatchRetryIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    @Before
+    public void setup() {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    private JobParameters defaultJobParameters() {
+        JobParametersBuilder paramsBuilder = new JobParametersBuilder();
+        paramsBuilder.addString("jobID", String.valueOf(System.currentTimeMillis()));
+        return paramsBuilder.toJobParameters();
+    }
```
Comment (ashleyfrieze) [inline spring-batch/src/test/java/org/baeldung/batch/SpringBatchRetryIntegrationTest.java]: Let's move this to after the `@Test` methods. The tests come first usually in a test class.

File: spring-batch/src/test/java/org/baeldung/batch/SpringBatchRetryIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+    @Mock
+    private CloseableHttpClient httpClient;
+
+    @Mock
+    private CloseableHttpResponse httpResponse;
+
+    @InjectMocks
+    @Autowired
+    private RetryItemProcessor retryItemProcessor;
```
Comment (ashleyfrieze) [inline spring-batch/src/test/java/org/baeldung/batch/SpringBatchRetryIntegrationTest.java]: This is a bit messy... we're injecting mocks AND also autowiring. Hard to follow. Does this version of spring support `@MockBean` for the `CloseableHttpClient`?

The problem you've got, though, is that the `RetryItemProcessor` creates its own http client. Really, the client should be injected via the constructor and should be defined in the `@Configuration` beans class. That would make this easier to follow.

The above does work. The `@Autowired` bean gets created by spring, injected into this class, and then line `66` does some hacking into the object to force the mocks in. It's essentially an anti-pattern for using these things together, especially as SpringTest supports mockito.

Can we tidy this up?

File: spring-batch/src/main/java/org/baeldung/batch/SpringBatchRetryConfig.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    @Value("input/recordRetry.csv")
+    private Resource inputCsv;
+
+    @Value("file:xml/retryOutput.xml")
+    private Resource outputXml;
+
+    public ItemReader<Transaction> itemReader(Resource inputData) throws UnexpectedInputException, ParseException {
+        FlatFileItemReader<Transaction> reader = new FlatFileItemReader<>();
+        DelimitedLineTokenizer tokenizer = new DelimitedLineTokenizer();
+        String[] tokens = { "username", "userid", "transactiondate", "amount" };
```
[the remark this answers is not in the document]
Comment (shubhi22) [reply spring-batch/src/main/java/org/baeldung/batch/SpringBatchRetryConfig.java]: done

File: spring-batch/src/main/java/org/baeldung/batch/SpringBatchRetryConfig.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        tokenizer.setNames(tokens);
+        reader.setResource(inputData);
```
[the remark this answers is not in the document]
Comment (shubhi22) [reply spring-batch/src/main/java/org/baeldung/batch/SpringBatchRetryConfig.java]: done

File: spring-batch/src/main/java/org/baeldung/batch/SpringBatchRetryConfig.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        DefaultLineMapper<Transaction> lineMapper = new DefaultLineMapper<>();
+        lineMapper.setLineTokenizer(tokenizer);
+        lineMapper.setFieldSetMapper(new RecordFieldSetMapper());
```
[the remark this answers is not in the document]
Comment (shubhi22) [reply spring-batch/src/main/java/org/baeldung/batch/SpringBatchRetryConfig.java]: done

File: spring-batch/src/main/java/org/baeldung/batch/SpringBatchRetryConfig.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            .retry(ConnectTimeoutException.class)
+            .retry(DeadlockLoserDataAccessException.class)
+            .build();
+    }
+
+    @Bean(name = "retryBatchJob")
+    public Job retryJob(
+        @Qualifier("retryStep")
+            Step retryStep) {
+        return jobBuilderFactory.get("retryBatchJob").start(retryStep).build();
```
[the remark this answers is not in the document]
Comment (shubhi22) [reply spring-batch/src/main/java/org/baeldung/batch/SpringBatchRetryConfig.java]: done

File: spring-batch/src/main/java/org/baeldung/batch/SpringBatchRetryConfig.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        Jaxb2Marshaller marshaller = new Jaxb2Marshaller();
+        marshaller.setClassesToBeBound(Transaction.class);
+        return marshaller;
+    }
+
+    @Bean
+    public Step retryStep(
+        @Qualifier("retryItemProcessor")
+            ItemProcessor<Transaction, Transaction> processor, ItemWriter<Transaction> writer) throws ParseException {
+        return stepBuilderFactory.get("retryStep").<Transaction, Transaction>chunk(10).reader(itemReader(inputCsv))
```
[the remark this answers is not in the document]
Comment (shubhi22) [reply spring-batch/src/main/java/org/baeldung/batch/SpringBatchRetryConfig.java]: done

File: spring-batch/src/main/java/org/baeldung/batch/service/RetryItemProcessor.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+import java.io.IOException;
+
+public class RetryItemProcessor implements ItemProcessor<Transaction, Transaction> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(RetryItemProcessor.class);
+
+    private CloseableHttpClient client;
+
+    public RetryItemProcessor() {
+        final RequestConfig config = RequestConfig.custom().setConnectTimeout(2 * 1000).build();
```
[the remark this answers is not in the document]
Comment (shubhi22) [reply spring-batch/src/main/java/org/baeldung/batch/service/RetryItemProcessor.java]: done

File: spring-batch/src/main/java/org/baeldung/batch/service/RetryItemProcessor.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        client = HttpClientBuilder.create().setDefaultRequestConfig(config).build();
+    }
+
+    @Override
+    public Transaction process(Transaction transaction) throws IOException, JSONException {
+        LOGGER.info("Attempting to process user with id={}", transaction.getUserId());
+        System.out.println("Attempting to process user with id=" + transaction.getUserId());
```
[the remark this answers is not in the document]
Comment (shubhi22) [reply spring-batch/src/main/java/org/baeldung/batch/service/RetryItemProcessor.java]: Had kept it considering the current logging level of this app is set to ERROR by default. And all other files (created as part of other articles) are using sysout for that reason.
Removed it though. Please let me know if you will like me to update the logback level to INFO

File: spring-batch/src/test/java/org/baeldung/batch/SpringBatchRetryIntegrationTest.java (hunk 1 above)
Comment (shubhi22) [reply spring-batch/src/test/java/org/baeldung/batch/SpringBatchRetryIntegrationTest.java]: done

File: spring-batch/src/test/java/org/baeldung/batch/SpringBatchRetryIntegrationTest.java (hunk 2 above)
Comment (shubhi22) [reply spring-batch/src/test/java/org/baeldung/batch/SpringBatchRetryIntegrationTest.java]: I had tried using @MockBean initially but was getting NullPointerException which I couldn't resolve at that moment so I went ahead with this approach.
However, I just realized it was due to some additional annotations used in this class for which it wasn't working(nullpointer). Resolved it and have now used @MockBean.
Please review if this looks good.
